// cause/blocks/noflo-component.js
'use strict';

require('coffee-script/register');
var _ = require('lodash');


function get_noflo_component(module_name) {
	return require(module_name).getComponent();
}


function fn(task, step, input, prev_step, done) {
	var cause = this;

	var component = get_noflo_component(step.options.component);
	var inport_name = step.options.inport_name;
	var outport_name = step.options.outport_name;

	component.compute = _.wrap(component.compute, function(compute) {
		// console.log('compute');
		compute();
	});

	// component.outPorts[outport_name].emit('attach');
	component.outPorts[outport_name].isAttached = function() {
		return true;
	};

	input = 'hsl(100, 50%, 75%)';
	// component.props.color = [input];
	// console.log(component.props);

	var output;
	component.outPorts[outport_name].send = _.wrap(
		component.outPorts[outport_name].send,
		function(send, payload) {
			output = payload;
			console.log(outport_name+':', output);
			// send();
		}
	);

	component.inPorts[inport_name].handleSocketEvent('data', input, 0);

	done(null, output, null);
}


module.exports = {
	fn: fn,
	defaults: {
		options: {
			component: 'noflo-color/components/isDark',
			inport_name: 'color',
			outport_name: 'dark'
		},
		data: {},
		description: "use noflo components with ’cause"
	},
};
